cmake/extensions.cmake

```diff
@@ -2261,3 +2263,368 @@ function(target_byproducts)
                      COMMENT "Logical command for additional byproducts on target: ${TB_TARGET}"
   )
 endfunction()
+
+########################################################
+# 4. Zephyr devicetree function
+########################################################
+# 4.1. dt_*
+#
+# The following methods are for retrieving devicetree information in CMake.
+#
+# Note: In CMake we refer to the nodes using the node's path, therefore there
+# is no dt_path(...) function for obtaining a node identifier.
+
+# Usage:
+#   dt_nodelabel(<var> LABEL <label>)
+#
+# function for retrieving the node path for the node having nodelabel
+# <label>.
+#
+# the node path will be returned in the <var> parameter.
+#
+# <var>          : Return variable where the node path will be stored
+# LABEL <label>  : Node label
+function(dt_nodelabel var)
+  set(req_single_args "LABEL")
+  cmake_parse_arguments(DT_LABEL "" "${req_single_args}" "" ${ARGN})
+
+  if(${ARGV0} IN_LIST req_single_args)
+    message(FATAL_ERROR "dt_nodelabel(${ARGV0} ...) missing return parameter.")
+  endif()
+
+  foreach(arg ${req_single_args})
+    if(NOT DEFINED DT_LABEL_${arg})
+      message(FATAL_ERROR "dt_nodelabel(${ARGV0} ...) "
+                          "missing required argument: ${arg}"
+      )
+    endif()
+  endforeach()
+
+  get_target_property(${var} devicetree_target "DT_LABEL|${DT_LABEL_LABEL}")
+  if(${${var}} STREQUAL ${var}-NOTFOUND)
+    set(${var})
```

Comment:
I kind of want this to be a fatal error at first, with a possibility to relax it later by adding a new `DEFAULT my-default-value` argument to req_single_args.

Is there a good reason not to do it that way?

The kconfigfunctions.py functions that return paths do return empty strings when there's no such node, but that is because we have to have *something* to return since we parse everything in the entire kconfig tree and sometimes we're asking about nodes that only exist on a different SoC. So I don't think they're a good example to follow here.

Reply:
to allow code to just ask for a node with a given label.

If no node is having that label, then there is nothing to return.

If someone asks for `some-node`, then how should they know prior to their call if there is any node with such label ?
And how will a fatal error help them.

By undefining the return var, they can do:
```
dt_label(node LABEL some-node)
if(DEFINED node)
  # Node exists, do something related to that node
endif()
```

Reply:
> If someone asks for `some-node`, then how should they know prior to their call if there is any node with such label ?

With a `dt_node_exists` helper, of course :)

Reply:
... or a `DEFAULT ""` extra keyword argument to the function, since the empty string evaluates to false.

Reply:
> With a `dt_node_exists` helper, of course :)

Could you explain a bit on how such a flow would look, cause i'm not completely sure I understand your intention.
I want to find the node with nodelabel "something", then how can I ask whether that node exists before getting it ?
```
dt_node_exists(result PATH ${node_path})  # <---  but I don't know the path, as that is what I will be asking for in next line
dt_label(node_path NODELABEL "something") # <-- Here I get the path.
```

Or did you mean that `dt_node_exists()` should also take node label as argument, like this:
```
dt_node_exists(result NODELABEL "something")
```

but in such case we'll end up with constructs like this:
```
dt_node_exists(result NODELABEL "something")
if(result)
  dt_label(node_path NODELABEL "something")
  # Now do something with that node, like
  # dt_prop(result PATH ${node_path})
endif()
```

but why is that better than:
```
dt_label(node_path NODELABEL "something")
if(DEFINED node_path)
  # Now do something with that node, like
  # dt_prop(result PATH ${node_path})
endif()
```
In CMake we can use the knowledge of whether the variable is undefined versus an empty string / TRUE / FALSE / value.

So in current PR, I have both the knowledge of whether or not the node exists, but also the knowledge of its node path if it does exists, so why split that into two function calls ?
What's the exact gain.

Reply:
> .. or a `DEFAULT ""` extra keyword argument to the function, since the empty string evaluates to false.

but we are not returning an empty string.
We are undefining the variable.

What would be the exact gain of returning some default when the node is not there ?

Reply:
Small example of the difference between undefined and an empty string:
```
set(TEST_VAR)

if(DEFINED TEST_VAR)
  message("TEST_VAR is defined, value=${TEST_VAR}")
else()
  message("TEST_VAR is undefined")
endif()

set(TEST_VAR "")

if(DEFINED TEST_VAR)
  message("TEST_VAR is defined, value=${TEST_VAR}")
else()
  message("TEST_VAR is undefined")
endif()
```
results in:
```
$ cmake -P z.cmake
TEST_VAR is undefined
TEST_VAR is defined, value=
```
So the `if(DEFINED <var>)` returns true, even for an empty string, but false if the variable is undefined.
```
set(MY_VAR)    # <-- This undefines the var
set(MY_VAR "") # <-- This defines the var to the empty string.
```

Reply:
OK, I am convinced. Thanks.